universal-application-tool-0.0.1/app/services/applicant/ApplicantServiceImpl.java

```diff
@@ -179,6 +181,104 @@ private void stageUpdates(
       UpdateMetadata updateMetadata,
       ImmutableSet<Update> updates)
       throws UnsupportedScalarTypeException, PathNotInBlockException {
+    if (block.isEnumerator()) {
+      stageEnumeratorUpdates(applicantData, block, updateMetadata, updates);
+    } else {
+      stageNormalUpdates(applicantData, block, updateMetadata, updates);
+    }
+  }
+
+  /**
+   * Stage updates for an enumerator.
+   *
+   * @throws PathNotInBlockException for updates that aren't {@link Scalar#ENTITY_NAME}, or {@link
+   *     Scalar#DELETE_ENTITY}.
+   */
+  private void stageEnumeratorUpdates(
+      ApplicantData applicantData,
+      Block block,
+      UpdateMetadata updateMetadata,
+      ImmutableSet<Update> updates)
+      throws PathNotInBlockException {
+    ImmutableSet<Update> addsAndChanges = validateEnumeratorAddsAndChanges(block, updates);
+    ImmutableSet<Update> deletes =
+        updates.stream()
+            .filter(
+                update ->
+                    update
+                        .path()
+                        .withoutArrayReference()
+                        .equals(Path.empty().join(Scalar.DELETE_ENTITY)))
+            .collect(ImmutableSet.toImmutableSet());
+
+    // If there are more updates than there are adds/changes and deletes, throw
+    Set<Update> unknownUpdates = Sets.difference(updates, Sets.union(addsAndChanges, deletes));
+    if (!unknownUpdates.isEmpty()) {
+      throw new PathNotInBlockException(block.getId(), unknownUpdates.iterator().next().path());
+    }
+
+    for (Update update : addsAndChanges) {
```

Comment:
nit: could we add a comment reminder that changes need to happen first,  before deletions possibly change the original index?

Reply:
done